File: test/points_test.py
import copclib as copc
import pytest
import random


def test_points_constructor():
    points = copc.Points(
        6, copc.Vector3.DefaultScale(), copc.Vector3.DefaultOffset(), 4
    )
    assert points.point_format_id == 6
    assert points.point_record_length == 34
    assert len(points) == 0

    point1 = copc.Points(
        6,
        copc.Vector3.DefaultScale(),
        copc.Vector3.DefaultOffset(),
        eb_byte_size=4,
    ).CreatePoint()
    point1.unscaled_x = 11
    point1.unscaled_y = 11
    point1.unscaled_z = 11

    point_list = [
        point1,
        copc.Points(
            6,
            copc.Vector3.DefaultScale(),
            copc.Vector3.DefaultOffset(),
            eb_byte_size=4,
        ).CreatePoint(),
        copc.Points(
            6,
            copc.Vector3.DefaultScale(),
            copc.Vector3.DefaultOffset(),
            eb_byte_size=4,
        ).CreatePoint(),
    ]

    points = copc.Points(point_list)
    assert points.point_format_id == 6
    assert points.point_record_length == 34
    for point in points:
        assert point.point_format_id == 6
    assert points[0].unscaled_y == 11
    assert points[0].unscaled_z == 11

    str(points)


def test_adding_point_to_points():
    points = copc.Points(
        6, copc.Vector3.DefaultScale(), copc.Vector3.DefaultOffset(), 0
    )
    point = copc.Points(
        6,
        copc.Vector3.DefaultScale(),
        copc.Vector3.DefaultOffset(),
        eb_byte_size=0,
    ).CreatePoint()
    point.unscaled_x = 11
    point.unscaled_y = 11
    point.unscaled_z = 11

    points.AddPoint(point)

    assert len(points) == 1
    assert points[0].unscaled_x == 11
    assert points[0].unscaled_y == 11
    assert points[0].unscaled_z == 11

    point = copc.Points(
        6,
        copc.Vector3.DefaultScale(),
        copc.Vector3.DefaultOffset(),
        eb_byte_size=0,
    ).CreatePoint()
    point.unscaled_x = 22
    point.unscaled_y = 22
    point.unscaled_z = 22

    points.AddPoint(point)
    assert len(points) == 2
    assert points[1].unscaled_x == 22
    assert points[1].unscaled_y == 22
    assert points[1].unscaled_z == 22

    # Test check on point format
    point = copc.Points(
        7,
        copc.Vector3.DefaultScale(),
        copc.Vector3.DefaultOffset(),
        eb_byte_size=0,
    ).CreatePoint()
    with pytest.raises(RuntimeError):
        points.AddPoint(point)

    # Test check on extra bytes
    point = copc.Points(
        6,
        copc.Vector3.DefaultScale(),
        copc.Vector3.DefaultOffset(),
        eb_byte_size=1,
    ).CreatePoint()
    with pytest.raises(RuntimeError):
        points.AddPoint(point)


def test_adding_points_to_points():
    points = copc.Points(
        [
            copc.Points(
                6,
                copc.Vector3.DefaultScale(),
                copc.Vector3.DefaultOffset(),
                eb_byte_size=4,
            ).CreatePoint()
            for _ in range(10)
        ]
    )
    points_other = copc.Points(
        [
            copc.Points(
                6,
                copc.Vector3.DefaultScale(),
                copc.Vector3.DefaultOffset(),
                eb_byte_size=4,
            ).CreatePoint()
            for _ in range(10)
        ]
    )

    points.AddPoints(points_other)

    assert len(points) == 20

    # Test check on point format
    points_other = copc.Points(
        [
            copc.Points(
                7,
                copc.Vector3.DefaultScale(),
                copc.Vector3.DefaultOffset(),
                eb_byte_size=4,
            ).CreatePoint()
            for _ in range(10)
        ]
    )
    with pytest.raises(RuntimeError):
        points.AddPoints(points_other)

    # Test check on extra bytes
    points_other = copc.Points(
        [
            copc.Points(
                6,
                copc.Vector3.DefaultScale(),
                copc.Vector3.DefaultOffset(),
                eb_byte_size=1,
            ).CreatePoint()
            for _ in range(10)
        ]
    )
    with pytest.raises(RuntimeError):
        points.AddPoints(points_other)


def test_points_format_conversion():
    points = copc.Points(
        [
            copc.Points(
                6,
                copc.Vector3.DefaultScale(),
                copc.Vector3.DefaultOffset(),
                eb_byte_size=4,
            ).CreatePoint()
            for _ in range(10)
        ]
    )
    points.ToPointFormat(7)

    assert points.point_format_id == 7
    assert points.point_record_length == 40
    assert points[0].point_format_id == 7
    assert points[0].point_record_length == 40

    with pytest.raises(RuntimeError):
        points.ToPointFormat(5)
    with pytest.raises(RuntimeError):
        points.ToPointFormat(9)


def test_points_iterator():
    points = copc.Points(
        6, copc.Vector3.DefaultScale(), copc.Vector3.DefaultOffset(), 4
    )

    # generate points
    num_points = 2000
    for i in range(num_points):
        p = points.CreatePoint()
        p.classification = i % 32
        points.AddPoint(p)

    classification_index = [points[i].classification for i in range(num_points)]
    classification_iterator = [p.classification for p in points]

    for i, clas in enumerate(classification_index):
        assert clas == i % 32

    assert classification_index == classification_iterator


def test_points_group_accessors():
    points = copc.Points(
        6, copc.Vector3.DefaultScale(), copc.Vector3.DefaultOffset(), 4
    )

    # generate points
    num_points = 2000
    for i in range(num_points):
        p = points.CreatePoint()
        p.x = i
        p.y = i * 3
        p.z = i - 80
        points.AddPoint(p)

    assert len(points) == num_points

    # test that the getters work
    for i in range(num_points):
        assert points.x[i] == i
        assert points.y[i] == i * 3
        assert points.z[i] == i - 80

    # generate vector of coordinates
    xn = []
    yn = []
    zn = []
    with pytest.raises(RuntimeError):
        points.x = xn
        points.y = yn
        points.z = zn

    for i in range(num_points - 1):
        xn.append(i * 50 + 8)
        yn.append(i + 800)
        zn.append(i * 4)

    with pytest.raises(RuntimeError):
        points.x = xn
        points.y = yn
        points.z = zn

    # add the last point
    xn.append(1)
    yn.append(2)
    zn.append(3)

    # test setters
    points.x = xn
    points.y = yn
    points.z = zn

    for i in range(num_points - 1):
        p = points[i]
        assert p.x == i * 50 + 8
        assert p.y == i + 800
        assert p.z == i * 4

    # test negative indices
    last_point = points[-1]
    assert last_point.x == 1
    assert last_point.y == 2
    assert last_point.z == 3


def test_within():

    points = copc.Points(6, (1, 1, 1), copc.Vector3.DefaultOffset())

    # generate points
    for i in range(2000):
        p = points.CreatePoint()
        p.x = random.uniform(0, 5)
        p.y = random.uniform(0, 5)
        p.z = random.uniform(0, 5)
        points.AddPoint(p)

    assert points.Within(copc.Box(0, 0, 0, 5, 5, 5))

    p = points.CreatePoint()
    p.x = random.uniform(0, 5)
    p.y = random.uniform(0, 5)
    p.z = 6
    points.AddPoint(p)

    assert not points.Within(copc.Box(0, 0, 0, 5, 5, 5))


def test_get_within():

    points = copc.Points(6, (1, 1, 1), copc.Vector3.DefaultOffset())

    # generate points
    for i in range(2000):
        p = points.CreatePoint()
        p.x = random.uniform(0, 5)
        p.y = random.uniform(0, 5)
        p.z = random.uniform(0, 5)
        points.AddPoint(p)

    box = copc.Box(0, 0, 0, 2.5, 2.5, 2.5)
    points = points.GetWithin(box)
    assert copc.Points(points).Within(box)


def test_points_accessors():
    points = copc.Points(
        7, copc.Vector3.DefaultScale(), copc.Vector3.DefaultOffset(), 4
    )

    # generate points
    num_points = 2000
    for i in range(num_points):
        p = points.CreatePoint()
        p.x = i
        p.y = i * 3
        p.z = i - 80
        p.classification = i * 255 // num_points
        p.point_source_id = i * 255 // num_points
        points.AddPoint(p)

    for i, (x, y, z, classification, point_source_id) in enumerate(
        zip(points.x, points.y, points.z, points.classification, points.point_source_id)
    ):
        assert x == i
        assert y == i * 3
        assert z == i - 80
        assert classification == i * 255 // num_points
        assert point_source_id == i * 255 // num_points

    # test slice
    assert points.x[5:10] == [x for x in range(5, 10)]
    assert points.y[-10:] == [x * 3 for x in range(1990, 2000)]
    assert points.z[:10] == [x - 80 for x in range(0, 10)]
    assert points.classification[:10] == [x * 255 // num_points for x in range(0, 10)]
    assert points.point_source_id[:10] == [x * 255 // num_points for x in range(0, 10)]

    # test index setter
    for i in range(len(points)):
        p = points[i]
        p.x = 20
        p.y = 30
        p.z = 40
        p.classification = 50
        p.point_source_id = 50

    assert all([x == 20 for x in points.x])
    assert all([y == 30 for y in points.y])
    assert all([z == 40 for z in points.z])
    assert all([classification == 50 for classification in points.classification])
    assert all([point_source_id == 50 for point_source_id in points.point_source_id])

    # test iterator setter
    for p in points:
        p.x = -50
        p.y = -60
        p.z = -70

    assert all([x == -50 for x in points.x])
    assert all([y == -60 for y in points.y])
    assert all([z == -70 for z in points.z])


def test_points_indexer_setter():
    points = copc.Points(
        6, copc.Vector3.DefaultScale(), copc.Vector3.DefaultOffset(), 4
    )

    # generate points
    num_points = 2000
    for i in range(num_points):
        points.AddPoint(points.CreatePoint())

    assert all([x == 0 for x in points.x])
    assert all([y == 0 for y in points.y])
    assert all([z == 0 for z in points.z])

    points[0].x = 20
    points[0].y = 40
    points[0].z = 80
    points[2].intensity = 60000

    assert points[0].x == 20
    assert points[0].y == 40
    assert points[0].z == 80
    assert points[2].intensity == 60000

    # test slicing setters
    points[:].x = [2] * len(points)
    points[:].y = [4] * len(points)
    points[:].z = [8] * len(points)

    assert all([x == 2 for x in points.x])
    assert all([x == 4 for x in points.y])
    assert all([x == 8 for x in points.z])

    # TODO: Allow the user to set all values the same
    points[1000:].x = [-2] * 1000
    points[1500:1600].y = [-4] * 100
    points[-5:].z = [-8] * 5

    assert all([x == -2 for i, x in enumerate(points.x) if i >= 1000])
    assert all([x != -2 for i, x in enumerate(points.x) if not i >= 1000])
    assert all([x == -4 for i, x in enumerate(points.y) if i >= 1500 and i < 1600])
    assert all(
        [x != -4 for i, x in enumerate(points.y) if not (i >= 1500 and i < 1600)]
    )
    assert all([x == -8 for i, x in enumerate(points.z) if i >= len(points) - 5])
    assert all([x != -8 for i, x in enumerate(points.z) if not (i >= len(points) - 5)])

File: test/points_test.cpp
#include <limits>
#include <random>

#include <catch2/catch.hpp>
#include <copc-lib/geometry/box.hpp>
#include <copc-lib/geometry/vector3.hpp>
#include <copc-lib/las/point.hpp>
#include <copc-lib/las/points.hpp>

using namespace copc::las;
using namespace std;

TEST_CASE("Points tests", "[Point]")
{
    SECTION("Points constructors")
    {
        auto points = Points(6, 4);
        REQUIRE(points.PointFormatId() == 6);
        REQUIRE(points.PointRecordLength() == 34);
        REQUIRE(points.Get().empty());

        auto point_vec = std::vector<std::shared_ptr<Point>>();
        auto point1 = points.CreatePoint();
        point1->X(11.1);
        point1->Y(11.2);
        point1->Z(11.3);
        point_vec.push_back(point1);

        auto point2 = points.CreatePoint();
        point_vec.push_back(point2);

        auto point3 = points.CreatePoint();
        point_vec.push_back(point3);

        points = Points(point_vec);
        REQUIRE(points.PointFormatId() == 6);
        REQUIRE(points.PointRecordLength() == 34);
        for (const auto &point : points.Get())
            REQUIRE(point->PointFormatId() == 6);
        REQUIRE(points.Get(0)->X() == 11.1);
        REQUIRE(points.Get(0)->Y() == 11.2);
        REQUIRE(points.Get(0)->Z() == 11.3);

        SECTION("Points ToString") 
        { 
            points.ToString();
            std::stringstream ss;
            ss << points;
        }    
    }

    SECTION("Adding Point to Points")
    {
        auto points = Points(6, 0);
        auto point = points.CreatePoint();
        point->X(11.1);
        point->Y(11.2);
        point->Z(11.3);

        points.AddPoint(point);

        REQUIRE(points.Get().size() == 1);
        REQUIRE(points.Get(0)->X() == 11.1);
        REQUIRE(points.Get(0)->Y() == 11.2);
        REQUIRE(points.Get(0)->Z() == 11.3);

        point = points.CreatePoint();
        point->X(22.5);
        point->Y(22.6);
        point->Z(22.7);

        points.AddPoint(point);
        REQUIRE(points.Get().size() == 2);
        REQUIRE(points.Get(1)->X() == 22.5);
        REQUIRE(points.Get(1)->Y() == 22.6);
        REQUIRE(points.Get(1)->Z() == 22.7);

        // Test check on point format
        point = std::make_shared<Point>(7, 0);
        REQUIRE_THROWS(points.AddPoint(point));

        // Test check on extra bytes
        point = std::make_shared<Point>(6, 1);
        REQUIRE_THROWS(points.AddPoint(point));
    }

    SECTION("Adding Points to Points")
    {
        auto points = Points(std::vector<std::shared_ptr<Point>>(10, std::make_shared<Point>(6, 4)));
        auto points_other = Points(std::vector<std::shared_ptr<Point>>(10, std::make_shared<Point>(6, 4)));

        points.AddPoints(points_other);

        REQUIRE(points.Get().size() == 20);

        // Test check on point format
        points_other = Points(std::vector<std::shared_ptr<Point>>(10, std::make_shared<Point>(7, 4)));
        REQUIRE_THROWS(points.AddPoints(points_other));

        // Test check on extra bytes
        points_other = Points(std::vector<std::shared_ptr<Point>>(10, std::make_shared<Point>(6, 1)));
        REQUIRE_THROWS(points.AddPoints(points_other));
    }

    SECTION("Points format conversion")
    {
        auto points = Points(std::vector<std::shared_ptr<Point>>(10, std::make_shared<Point>(6, 4)));
        points.ToPointFormat(7);

        REQUIRE(points.PointFormatId() == 7);
        REQUIRE(points.PointRecordLength() == 40);
        REQUIRE(points.Get(1)->PointFormatId() == 7);
        REQUIRE(points.Get(1)->PointRecordLength() == 40);

        REQUIRE_THROWS(points.ToPointFormat(5));
        REQUIRE_THROWS(points.ToPointFormat(9));
    }

    SECTION("Points Group Accessors")
    {
        auto points = Points(7, 4);

        // generate points
        int num_points = 2000;
        for (int i = 0; i < num_points; i++)
        {
            auto p = points.CreatePoint();
            p->X(i);
            p->Y(i * 3);
            p->Z(i - 80);
            p->Classification(i * 255 / num_points);
            p->PointSourceId(i * 255 / num_points);
            p->Red(i * 4);
            p->Green(i * 5);
            p->Blue(i * 6);
            points.AddPoint(p);
        }

        REQUIRE(points.Size() == num_points);

        // test that the getters work
        auto X = points.X();
        auto Y = points.Y();
        auto Z = points.Z();
        auto classification = points.Classification();
        auto point_source_id = points.PointSourceId();
        auto red = points.Red();
        auto green = points.Green();
        auto blue = points.Blue();
        for (int i = 0; i < num_points; i++)
        {
            REQUIRE(X[i] == i);
            REQUIRE(Y[i] == i * 3);
            REQUIRE(Z[i] == i - 80);
            REQUIRE(classification[i] == i * 255 / num_points);
            REQUIRE(point_source_id[i] == i * 255 / num_points);
            REQUIRE(red[i] == i * 4);
            REQUIRE(green[i] == i * 5);
            REQUIRE(blue[i] == i * 6);
        }

        // generate vector of coordinates
        std::vector<double> Xn;
        std::vector<double> Yn;
        std::vector<double> Zn;
        std::vector<int32_t> UnscaledXn;
        std::vector<int32_t> UnscaledYn;
        std::vector<int32_t> UnscaledZn;
        std::vector<uint8_t> classification_n;
        std::vector<uint8_t> point_source_id_n;
        std::vector<uint16_t> red_n;
        std::vector<uint16_t> green_n;
        std::vector<uint16_t> blue_n;

        REQUIRE_THROWS(points.X(Xn));
        REQUIRE_THROWS(points.Y(Yn));
        REQUIRE_THROWS(points.Z(Zn));
        REQUIRE_THROWS(points.Classification(classification_n));
        REQUIRE_THROWS(points.PointSourceId(point_source_id_n));
        REQUIRE_THROWS(points.Red(red_n));
        REQUIRE_THROWS(points.Green(green_n));
        REQUIRE_THROWS(points.Blue(blue_n));

        for (int i = 0; i < num_points - 1; i++)
        {
            Xn.push_back(i * 50 + 8);
            Yn.push_back(i + 800);
            Zn.push_back(i * 4);
            classification_n.push_back(i * 255 / 2000);
            point_source_id_n.push_back(i * 255 / 2000);
            red_n.push_back(i * 4);
            green_n.push_back(i * 5);
            blue_n.push_back(i * 6);
        }

        REQUIRE_THROWS(points.X(Xn));
        REQUIRE_THROWS(points.Y(Yn));
        REQUIRE_THROWS(points.Z(Zn));
        REQUIRE_THROWS(points.Classification(classification_n));
        REQUIRE_THROWS(points.PointSourceId(point_source_id_n));
        REQUIRE_THROWS(points.Red(red_n));
        REQUIRE_THROWS(points.Green(green_n));
        REQUIRE_THROWS(points.Blue(blue_n));

        // add the last point
        Xn.push_back(1);
        Yn.push_back(2);
        Zn.push_back(3);
        classification_n.push_back(255);
        point_source_id_n.push_back(255);
        red_n.push_back(num_points * 4);
        green_n.push_back(num_points * 5);
        blue_n.push_back(num_points * 6);

        // test setters
        REQUIRE_NOTHROW(points.X(Xn));
        REQUIRE_NOTHROW(points.Y(Yn));
        REQUIRE_NOTHROW(points.Z(Zn));
        REQUIRE_NOTHROW(points.Classification(classification_n));
        REQUIRE_NOTHROW(points.PointSourceId(point_source_id_n));
        REQUIRE_NOTHROW(points.Red(red_n));
        REQUIRE_NOTHROW(points.Green(green_n));
        REQUIRE_NOTHROW(points.Blue(blue_n));

        for (int i = 0; i < num_points - 1; i++)
        {
            auto p = points.Get(i);
            REQUIRE(p->X() == i * 50 + 8);
            REQUIRE(p->Y() == i + 800);
            REQUIRE(p->Z() == i * 4);
            REQUIRE(p->Classification() == i * 255 / 2000);
            REQUIRE(p->PointSourceId() == i * 255 / 2000);
            REQUIRE(p->Red() == i * 4);
            REQUIRE(p->Green() == i * 5);
            REQUIRE(p->Blue() == i * 6);
        }

        // test last point
        auto last_point = points.Get(points.Size() - 1);
        REQUIRE(last_point->X() == 1);
        REQUIRE(last_point->Y() == 2);
        REQUIRE(last_point->Z() == 3);
        REQUIRE(last_point->Classification() == 255);
        REQUIRE(last_point->PointSourceId() == 255);
        REQUIRE(last_point->Red() == num_points * 4);
        REQUIRE(last_point->Green() == num_points * 5);
        REQUIRE(last_point->Blue() == num_points * 6);
    }

    SECTION("Points Indexers")
    {
        auto points = Points(6, 4);

        // generate points
        int num_points = 2000;
        for (int i = 0; i < num_points; i++)
        {
            auto p = points.CreatePoint();
            p->X(i);
            p->Y(i * 3);
            p->Z(i - 80);
            points.AddPoint(p);
        }

        REQUIRE(points[0]->X() == 0);
        REQUIRE(points.Get(0)->X() == 0);
        REQUIRE(points[0]->Y() == 0);
        REQUIRE(points.Get(0)->Y() == 0);
        REQUIRE(points[0]->Z() == 0 - 80);
        REQUIRE(points.Get(0)->Z() == 0 - 80);

        REQUIRE(points[1999]->X() == 1999);
        REQUIRE(points.Get(1999)->X() == 1999);
        REQUIRE(points[1999]->Y() == 1999 * 3);
        REQUIRE(points.Get(1999)->Y() == 1999 * 3);
        REQUIRE(points[1999]->Z() == 1999 - 80);
        REQUIRE(points.Get(1999)->Z() == 1999 - 80);

        for (int i = 0; i < num_points; i++)
        {
            points[i]->X(-i);
            points[i]->Y(-i * 2);
            points[i]->Z(-i * 3);
        }

        REQUIRE(points[0]->X() == 0);
        REQUIRE(points.Get(0)->X() == 0);
        REQUIRE(points[0]->Y() == 0);
        REQUIRE(points.Get(0)->Y() == 0);
        REQUIRE(points[0]->Z() == 0);
        REQUIRE(points.Get(0)->Z() == 0);

        REQUIRE(points[1999]->X() == -1999);
        REQUIRE(points.Get(1999)->X() == -1999);
        REQUIRE(points[1999]->Y() == -1999 * 2);
        REQUIRE(points.Get(1999)->Y() == -1999 * 2);
        REQUIRE(points[1999]->Z() == -1999 * 3);
        REQUIRE(points.Get(1999)->Z() == -1999 * 3);

        // generate vector of coordinates
        std::vector<double> Xn;
        std::vector<double> Yn;
        std::vector<double> Zn;

        REQUIRE_THROWS(points.X(Xn));
        REQUIRE_THROWS(points.Y(Yn));
        REQUIRE_THROWS(points.Z(Zn));

        for (int i = 0; i < num_points - 1; i++)
        {
            Xn.push_back(i * 50 + 8);
            Yn.push_back(i + 800);
            Zn.push_back(i * 4);
        }

        REQUIRE_THROWS(points.X(Xn));
        REQUIRE_THROWS(points.Y(Yn));
        REQUIRE_THROWS(points.Z(Zn));

        // add the last point
        Xn.push_back(1);
        Yn.push_back(2);
        Zn.push_back(3);

        // test setters
        REQUIRE_NOTHROW(points.X(Xn));
        REQUIRE_NOTHROW(points.Y(Yn));
        REQUIRE_NOTHROW(points.Z(Zn));

        for (int i = 0; i < num_points - 1; i++)
        {
            auto p = points.Get(i);
            REQUIRE(p->X() == i * 50 + 8);
            REQUIRE(p->Y() == i + 800);
            REQUIRE(p->Z() == i * 4);
        }

        // test last point
        auto last_point = points.Get(points.Size() - 1);
        REQUIRE(last_point->X() == 1);
        REQUIRE(last_point->Y() == 2);
        REQUIRE(last_point->Z() == 3);
    }

    SECTION("Within")
    {
        auto points = Points(6);

        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<double> dist(0.0, 5.0);

        // generate points
        int num_points = 2000;
        for (int i = 0; i < num_points; i++)
        {
            auto p = points.CreatePoint();
            p->X(dist(gen));
            p->Y(dist(gen));
            p->Z(dist(gen));
            points.AddPoint(p);
        }

        REQUIRE(points.Within(copc::Box(0, 0, 0, 5, 5, 5)));

        auto p = points.CreatePoint();
        p->X(dist(gen));
        p->Y(dist(gen));
        p->Z(6);
        points.AddPoint(p);

        REQUIRE(!points.Within(copc::Box(0, 0, 0, 5, 5, 5)));
    }

    SECTION("GetWithin")
    {
        auto points = Points(6);

        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<double> dist(0.0, 5.0);

        // generate points
        int num_points = 2000;
        for (int i = 0; i < num_points; i++)
        {
            auto p = points.CreatePoint();
            p->X(dist(gen));
            p->Y(dist(gen));
            p->Z(dist(gen));
            points.AddPoint(p);
        }
        auto box = copc::Box(0, 0, 0, 2.5, 2.5, 2.5);
        points = points.GetWithin(box);

        REQUIRE(points.Within(box));
    }
}
